src/akkudoktoreos/utils/visualize.py

```diff
@@ -0,0 +1,497 @@
+    )
+    report.finalize_group()  # Finalize the second group of charts
+
+    # Group 3: Adding a violin plot
+    data = [np.random.normal(0, std, 100) for std in range(1, 5)]  # Example data for violin plot
+    report.create_violin_plot(
+        data,
+        labels=["Group 1", "Group 2", "Group 3", "Group 4"],
+        title="Violin Plot",
+        xlabel="Groups",
+        ylabel="Values",
+    )
+    data = [np.random.normal(0, 1, 100)]  # Example data for violin plot
+    report.create_violin_plot(
+        data, labels=["Group 1"], title="Violin Plot", xlabel="Group", ylabel="Values"
+    )
+
+    report.finalize_group()  # Finalize the third group of charts
+
+    # Generate the PDF report
+    report.generate_pdf()
+
+
+from akkudoktoreos.optimization.genetic import OptimizationParameters  # circular import
+
+
+def prepare_visualize(
+    parameters: OptimizationParameters,
+    results: dict,
+    filename: str = "visualization_results_new.pdf",
+) -> None:
+    report = VisualizationReport(filename)
+    # Group 1:
+    report.create_line_chart(
+        None,
+        [parameters.ems.gesamtlast],
+        title="Load Profile",
+        xlabel="Hours",
+        ylabel="Load (Wh)",
+        labels=["Total Load (Wh)"],
+        markers=["s"],
+        line_styles=["-"],
+    )
+    report.create_line_chart(
+        None,
+        [parameters.ems.pv_prognose_wh],
+        title="PV Forecast",
+        xlabel="Hours",
+        ylabel="PV Generation (Wh)",
+    )
+
+    report.create_line_chart(
+        None,
+        [np.full(len(parameters.ems.gesamtlast), parameters.ems.einspeiseverguetung_euro_pro_wh)],
+        title="Remuneration",
+        xlabel="Hours",
+        ylabel="€/Wh",
+    )
+    if parameters.temperature_forecast:
+        report.create_line_chart(
+            None,
+            [parameters.temperature_forecast],
+            title="Temperature Forecast",
+            xlabel="Hours",
+            ylabel="°C",
+        )
+    report.finalize_group()
+
+    # Group 2:
+    report.create_line_chart(
+        None,
+        [
+            results["result"]["Last_Wh_pro_Stunde"],
+            results["result"]["Home_appliance_wh_per_hour"],
+            results["result"]["Netzeinspeisung_Wh_pro_Stunde"],
+            results["result"]["Netzbezug_Wh_pro_Stunde"],
+            results["result"]["Verluste_Pro_Stunde"],
+        ],
+        title="Energy Flow per Hour",
+        xlabel="Hours",
+        ylabel="Energy (Wh)",
+        labels=[
+            "Load (Wh)",
+            "Household Device (Wh)",
+            "Grid Feed-in (Wh)",
+            "Grid Consumption (Wh)",
+            "Losses (Wh)",
+        ],
+        markers=["o", "o", "x", "^", "^"],
+        line_styles=["-", "--", ":", "-.", "-"],
+    )
+    report.finalize_group()
+
+    # Group 3:
+    report.create_line_chart(
+        None,
+        [results["result"]["akku_soc_pro_stunde"], results["result"]["EAuto_SoC_pro_Stunde"]],
```

Comment:
add legend (labels?) for those. Can eauto_soc_pro_stunde be none?

Reply:
added